### PR TITLE
New resource azurerm_resource_provider

### LINKED ISSUE #1832: Resource to Enable Azure Providers
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

It would be great if we had a resource akin to [Google Project Service](https://www.terraform.io/docs/providers/google/r/google_project_service.html) which would allow us to turn on the Azure providers in code.  This would allow modules to declaratively state which providers they need enabled.

### New or Affected Resource(s)

* azurerm_provider_service

### Potential Terraform Configuration

```hcl
resource "azurerm_provider_service" "providers" {
  subscription_id = "your-subscription-id"
  tenant_id = "your-tenant-id"
  provider_namespaces = ["Microsoft.Example", "Microsoft.Foo", "Microsoft.Bar"]
}
```

### References

* https://docs.microsoft.com/en-us/rest/api/resources/providers/register

### PR DESCRIPTION
**Pull request by Humoiz**
Implementing feature mentioned in the feature request: #1832

This PR adds following new resource, tests and its documentation:
azurerm_resource_provider_registration

It allows to register specified resource providers.

(fixes #1832)

### CODE REVIEW

File: website/docs/r/resource_provider.html.markdown
Comment (ArcturusZhang) [inline website/docs/r/resource_provider.html.markdown:4]: How about the `azurerm_resource_provider_registration` for this resource name? I suppose this rsource should manages the registration of a resource provider rather than manages a resource provider. What do you think?

File: website/docs/r/resource_provider.html.markdown
Comment (ArcturusZhang) [inline website/docs/r/resource_provider.html.markdown:17]: Seems this data source is not used, could we remove this here?

File: website/docs/r/resource_provider.html.markdown
Comment (Humoiz) [reply website/docs/r/resource_provider.html.markdown:4]: @ArcturusZhang - Thanks for your feedback. I agree with all your comments, will address it in the next iteration.

**Comment by ArcturusZhang** [comment]
Fixes #1832

File: website/docs/r/resource_provider_registration.html.markdown
Comment (katbyte) [inline website/docs/r/resource_provider_registration.html.markdown:11]: We might want to point out that this is only applicable right now when the provider option to skip registration is configured

File: azurerm/internal/services/resource/resource_arm_provider_registration.go
Comment (Humoiz) [reply azurerm/internal/services/resource/resource_arm_provider_registration.go:36]: I would say the user should implement the for_each on the resource in case of multiple RPs.

variable "provider_namespaces" {
  type    = list(string)
  default = ["Microsoft.Management"]
}

resource "azurerm_resource_provider_registration" "example" {
  for_each = toset(var.provider_namespaces)

  resource_provider_namespace = each.value
}

The underlying API does not support taking the list of RPs as well.

File: website/docs/r/resource_provider_registration.html.markdown
Comment (Humoiz) [reply website/docs/r/resource_provider_registration.html.markdown:11]: I`ll add that to the next iteration.

File: azurerm/internal/services/resource/resource_arm_provider_registration.go
Comment (Humoiz) [reply azurerm/internal/services/resource/resource_arm_provider_registration.go:146]: Just to clarify, are you saying not to unregister the RP in case of delete?

File: azurerm/internal/services/resource/resource_arm_provider_registration.go
Comment (WodansSon) [reply azurerm/internal/services/resource/resource_arm_provider_registration.go:36]: @Humoiz, I think what @katbyte is asking for here is something like this:

```
"resource_providers": {
	Type:     schema.TypeList,
	Required: true,
	MaxItems: 50,
	Elem: &schema.Schema{
		Type: schema.TypeString,
		ValidateFunc: validation.<YourCustomValidationFunctionHere>,
	},
},
```
That way when you define the resource you can pass the resource providers like this:
```
resource "azurerm_resource_provider_registration" "example" {
  resource_providers = ["Microsoft.Management", "Microsoft.Network", "Microsoft.Storage"]
}
```
In your create function you will have to loop over the names in the slice and call the API individually for each of the resource providers that the user wants to have registered. Personally, I would do a get on the provider and check to see if it's registered or not, if it is not call the API to register it, if there is a failure add the error to a error slice and then move to the next provider that needs to be registered. Once you have walked the entire list of resource providers check to see if the error slice is empty or not, if not return the error slice else all registration calls were successful call the read function and exist with a success. Does that make sense?

Comment (tombuildsstuff) [reply azurerm/internal/services/resource/resource_arm_provider_registration.go:36]: I'd disagree here - as @Humoiz has mentioned we can use a foreach on the resource to manage individual resources

From Terraform's perspective it's better to have a resource per RP - particularly when longer term there's features which can be enabled/disabled here - which could make sense within this resource; so I think this is fine as-is 👍

File: azurerm/internal/services/resource/tests/resource_arm_provider_registration_test.go
Comment (tombuildsstuff) [inline azurerm/internal/services/resource/tests/resource_arm_provider_registration_test.go:28]: can we also add an import check here to check this imports correctly:

```suggestion
			},
			data.ImportStep(),
```

File: website/docs/r/resource_provider_registration.html.markdown
Comment (tombuildsstuff) [inline website/docs/r/resource_provider_registration.html.markdown:25]: can we make this:

```suggestion
* `name` - (Required) The namespace of the Resource Provider which should be registered. Changing this forces a new resource to be created.
```

File: website/docs/r/resource_provider_registration.html.markdown
Comment (tombuildsstuff) [inline website/docs/r/resource_provider_registration.html.markdown:34]: this resource also supports Import - can we include an example of how to do that ([example](https://github.com/terraform-providers/terraform-provider-azurerm/blame/master/website/docs/r/resource_group.html.markdown#L49-L55))

File: website/docs/r/resource_provider_registration.html.markdown
Comment (tombuildsstuff) [inline website/docs/r/resource_provider_registration.html.markdown:32]:
```suggestion
```

File: website/docs/r/resource_provider_registration.html.markdown
Comment (tombuildsstuff) [inline website/docs/r/resource_provider_registration.html.markdown:17]: (as above) this field wants to be `name`:

```suggestion
  name = "Microsoft.PolicyInsights"
```

File: azurerm/internal/services/resource/tests/resource_arm_provider_registration_test.go
Comment (tombuildsstuff) [inline azurerm/internal/services/resource/tests/resource_arm_provider_registration_test.go:96]: (as above) this wants to be `name`:

```suggestion
  name = "Microsoft.PolicyInsights"
```

File: azurerm/internal/services/resource/tests/resource_arm_provider_registration_test.go
Comment (tombuildsstuff) [inline azurerm/internal/services/resource/tests/resource_arm_provider_registration_test.go:43]: (as above) this wants to be `name`:

```suggestion
		resourceProviderNamespace := rs.Primary.Attributes["name"]
```

File: azurerm/internal/services/resource/resource_arm_provider_registration.go
Comment (tombuildsstuff) [inline azurerm/internal/services/resource/resource_arm_provider_registration.go:49]: (as above) this wants to be `name` here:

```suggestion
	resourceProviderNamespace := d.Get("name").(string)
```

File: azurerm/internal/services/resource/resource_arm_provider_registration.go
Comment (tombuildsstuff) [inline azurerm/internal/services/resource/resource_arm_provider_registration.go:36]: this wants to be `name` rather than `resource_provider_namespace` - whilst `namespace` is technically correct with regard to this API - every other resource uses `name` - so we may as well make this consistent here:

```suggestion
			"name": {
```

File: azurerm/internal/services/resource/resource_arm_provider_registration.go
Comment (tombuildsstuff) [reply azurerm/internal/services/resource/resource_arm_provider_registration.go:38]: it'd be good to add validation here - for now that'd be:

```suggestion
				Required: true,
				ValidateFunc:  validate.NoEmptyStrings,
```

but it'd be good to add a TODO to add enhanced validation to confirm this RP is available in Azure Cloud being used at plan time

File: azurerm/internal/services/resource/resource_arm_provider_registration.go
Comment (tombuildsstuff) [inline azurerm/internal/services/resource/resource_arm_provider_registration.go:46]: _minor_ since this is now Create-only rather than Create-Update we can make this:

```suggestion
	ctx, cancel := timeouts.ForCreate(meta.(*clients.Client).StopContext, d)
```

File: azurerm/internal/services/resource/resource_arm_provider_registration.go
Comment (tombuildsstuff) [inline azurerm/internal/services/resource/resource_arm_provider_registration.go:51]: we'll also probably need to add a check to ensure that automatic Resource Provider registration is disabled here - since otherwise Terraform and this resource provider will conflict, which isn't ideal

that information isn't exposed outside the initial provider block right now - so that may need to be threaded through [from here](https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/internal/clients/builder.go#L20) into [the account](https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/internal/clients/builder.go#L62) where it should then be available as:

```
meta.(*clients.Client).Account.SkipResourceProviderRegistration
```

File: azurerm/internal/services/resource/tests/resource_arm_provider_registration_test.go
Comment (tombuildsstuff) [inline azurerm/internal/services/resource/tests/resource_arm_provider_registration_test.go:32]: can we add a matching 'requires import' test here too to go with the new functionality above? [example](https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/internal/services/resource/tests/resource_arm_resource_group_test.go#L33-L50)

**Review by tombuildsstuff — DISMISSED** [review]
hey @Humoiz 

Thanks for pushing those changes - I've taken a look through and left some comments inline but if we can fix those up (and the tests pass) then this otherwise looks good to me 👍 

Thanks!

File: azurerm/internal/services/resource/resource_arm_provider_registration.go
Comment (grayzu) [reply azurerm/internal/services/resource/resource_arm_provider_registration.go:51]: @tombuildsstuff, this makes sense but what about the scenario/situation where we are not automatically registering a resource and the automatic registration is not disabled?

Comment (tombuildsstuff) [reply azurerm/internal/services/resource/resource_arm_provider_registration.go:51]: @grayzu good question - in practice we'll continue to add to the automatic RP registration list over time as new services appear (whether this is defaulted to registering them by default/not is a separate question) - so it's probably safer to check this regardless, rather than returning an error later should an RP get added to the "auto register" list?

**Comment by Humoiz** [comment]
@tombuildsstuff Thanks for the feedback. I`ll review it and address it soon.

**Comment by tombuildsstuff** [comment]
hey @Humoiz 

Apologies for the delayed response here - thought I'd replied to this - but thanks for pushing those changes.

Chatting with @grayzu last week, we think there's 4 scenarios we need to cover:

- if auto-registration is enabled, and a user tries to enable an RP (which we would auto-register) - error
- if auto-registration is enabled, and a user tries to register an RP we're not auto-registering (that's not already registered) - allow/register
- if auto-registration is disabled, and a user tries to register an RP that's already registered - raise a "requires import" error
- if auto-registration is disabled, and a user tries to register an RP that's not already registered - allow/register

Does that make sense?

**Comment by Humoiz** [comment]
> hey @Humoiz
> 
> Apologies for the delayed response here - thought I'd replied to this - but thanks for pushing those changes.
> 
> Chatting with @grayzu last week, we think there's 4 scenarios we need to cover:
> 
> * if auto-registration is enabled, and a user tries to enable an RP (which we would auto-register) - error
> * if auto-registration is enabled, and a user tries to register an RP we're not auto-registering (that's not already registered) - allow/register
> * if auto-registration is disabled, and a user tries to register an RP that's already registered - raise a "requires import" error
> * if auto-registration is disabled, and a user tries to register an RP that's not already registered - allow/register
> 
> Does that make sense?

@tombuildsstuff - For the scenario, where auto-registration is enabled, and a user tries to register an RP we're not auto-registering -allow register. How do I determine the ones which terraform is not registering?

**Comment by tombuildsstuff** [comment]
@Humoiz at the moment that's a little tricky since you'll have a circular reference - I'll refactor this tomorrow so that's not the case but in theory you'd get [this list](https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/internal/provider/required_resource_providers.go#L16) and string compare them (but as mentioned before, that's not possible right now due to a circular reference).

I'll post an update in the morning once that's been moved which'll make this possible - sorry for the delay.

**Comment by tombuildsstuff** [comment]
Opened #8309 which'll unblock this - once that's merged (and this PR is rebased on top of master) - it should be possible to do:

```
name := "Microsoft.Foo" // assuming being parsed from the config
for _, resourceProvider := range resourceproviders.Required() {
  if strings.EqualsFold(resourceProvider, name) {
    return fmt.Errorf(`The Resource Provider %q is automatically registered by Terraform.

To manage this Resource Provider Registration with Terraform you need to opt-out of Automatic Resource Provider Registration (by setting 'skip_provider_registration' to 'true' in the Provider block) to avoid conflicting with Terraform`, name)
  }
}
```

This is a little different to most other resources since this'll need to be in the Create, Read and Delete functions - as since this value can change between runs we need this to be consistent

**Comment by Humoiz** [comment]
I`ll rebase this PR and make the necessary changes soon!

**Comment by WodansSon** [comment]
@Humoiz, just curious if you were still going to follow up on the suggested changes by @tombuildsstuff ? If not that's fine, but we would like to build on top of what you have provided thus far so we can ship it in the next release of the provider. Do you mind if we step in and finish the resource up? Thanks again for the PR 🚀

**Comment by Humoiz** [comment]
> @Humoiz, just curious if you were still going to follow up on the suggested changes by @tombuildsstuff ? If not that's fine, but we would like to build on top of what you have provided thus far so we can ship it in the next release of the provider. Do you mind if we step in and finish the resource up? Thanks again for the PR 🚀

I am side tracked with some other things but I`m planning to get to it by this weekend. If you can fix it before than thats great otherwise I`ll finish it up.

**Comment by Humoiz** [comment]
> > @Humoiz, just curious if you were still going to follow up on the suggested changes by @tombuildsstuff ? If not that's fine, but we would like to build on top of what you have provided thus far so we can ship it in the next release of the provider. Do you mind if we step in and finish the resource up? Thanks again for the PR 🚀
> 
@WodansSon  I am side tracked with some other things but I`m planning to get to it by this weekend. If you can fix it before than thats great otherwise I`ll finish it up.

**Comment by Humoiz** [comment]
> Opened #8309 which'll unblock this - once that's merged (and this PR is rebased on top of master) - it should be possible to do:
> 
> ```
> name := "Microsoft.Foo" // assuming being parsed from the config
> for _, resourceProvider := range resourceproviders.Required() {
>   if strings.EqualsFold(resourceProvider, name) {
>     return fmt.Errorf(`The Resource Provider %q is automatically registered by Terraform.
> 
> To manage this Resource Provider Registration with Terraform you need to opt-out of Automatic Resource Provider Registration (by setting 'skip_provider_registration' to 'true' in the Provider block) to avoid conflicting with Terraform`, name)
>   }
> }
> ```
> 
> This is a little different to most other resources since this'll need to be in the Create, Read and Delete functions - as since this value can change between runs we need this to be consistent

@tombuildsstuff - Apologies for the delayed response. I have fixed all 4 scenarios where auto-registration is enabled/disabled. Please review and let me know what do you think. The tests are also passing.

cc: @WodansSon

**Review by katbyte — DISMISSED** [review]
Hi @Humoiz,

I just ran the tests for this PR and it seems like they are failing:
```
TestAccAzureRMResourceProviderRegistration_requiresImport: testing.go:684: Step 0 error: errors during apply:
        
        Error: A resource with the ID "/subscriptions/*******/providers/Microsoft.PolicyInsights" already exists - to be managed via Terraform this resource needs to be imported into the State. Please see the resource documentation for "azurerm_resource_provider_registration" for more information.
        
          on /opt/teamcity-agent/temp/buildTmp/tf-test205268533/main.tf line 7:
          (source code not available)
        
        
    TestAccAzureRMResourceProviderRegistration_basic: testing.go:684: Step 0 error: errors during apply:
        
        Error: A resource with the ID "/subscriptions/*******/providers/Microsoft.PolicyInsights" already exists - to be managed via Terraform this resource needs to be imported into the State. Please see the resource documentation for "azurerm_resource_provider_registration" for more information.
        
          on /opt/teamcity-agent/temp/buildTmp/tf-test158967470/main.tf line 7:
          (source code not available)
```

**Comment by Humoiz** [comment]
> Hi @Humoiz,
> 
> I just ran the tests for this PR and it seems like they are failing:
> 
> ```
> TestAccAzureRMResourceProviderRegistration_requiresImport: testing.go:684: Step 0 error: errors during apply:
>         
>         Error: A resource with the ID "/subscriptions/*******/providers/Microsoft.PolicyInsights" already exists - to be managed via Terraform this resource needs to be imported into the State. Please see the resource documentation for "azurerm_resource_provider_registration" for more information.
>         
>           on /opt/teamcity-agent/temp/buildTmp/tf-test205268533/main.tf line 7:
>           (source code not available)
>         
>         
>     TestAccAzureRMResourceProviderRegistration_basic: testing.go:684: Step 0 error: errors during apply:
>         
>         Error: A resource with the ID "/subscriptions/*******/providers/Microsoft.PolicyInsights" already exists - to be managed via Terraform this resource needs to be imported into the State. Please see the resource documentation for "azurerm_resource_provider_registration" for more information.
>         
>           on /opt/teamcity-agent/temp/buildTmp/tf-test158967470/main.tf line 7:
>           (source code not available)
> ```
Ok so these tests pass for me locally. Microsoft.PolicyInsights provider is not registered in my subscription and it may be registered in the environment where you are running the tests. Which Provider can I use for the tests?

File: azurerm/internal/services/resource/resource_arm_provider_registration.go
Comment (WodansSon) [inline azurerm/internal/services/resource/resource_arm_provider_registration.go:76]: Could we maybe have a different error message here, I feel if we display the `ImportAsExistsError` this may confuse end users, WDYT?

Import Error:
```
        Error: A resource with the ID "/subscriptions/*******/providers/Microsoft.PolicyInsights" already exists - to be managed via Terraform this resource needs to be imported into the State. Please see the resource documentation for "azurerm_resource_provider_registration" for more information.
        
          on /opt/teamcity-agent/temp/buildTmp/tf-test205268533/main.tf line 7:
          (source code not available)
```

Possible New Error:
```
        Error: The "/subscriptions/*******/providers/Microsoft.PolicyInsights" provider is currently already registered in this subscription.
        
          on /opt/teamcity-agent/temp/buildTmp/tf-test205268533/main.tf line 7:
          (source code not available)
```

**Review by WodansSon — APPROVED** [review]
@Humoiz, thanks for pushing these changes... I had one minor comment but other than that this LGTM! 🚀

File: azurerm/internal/services/resource/resource_arm_provider_registration.go
Comment (Humoiz) [reply azurerm/internal/services/resource/resource_arm_provider_registration.go:76]: I think the user should still be notified that this needs to imported in to the state. I can change it to 

The "/subscriptions/*******/providers/Microsoft.PolicyInsights" provider is currently already registered in this subscription. To be managed via Terraform this resource needs to be imported into the State. Please see the resource documentation for "azurerm_resource_provider_registration" for more information.


@tombuildsstuff @WodansSon - WDYT?

Comment (WodansSon) [reply azurerm/internal/services/resource/resource_arm_provider_registration.go:76]: @Humoiz , that works for me! 💯

Comment (WodansSon) [reply azurerm/internal/services/resource/resource_arm_provider_registration.go:76]: Personally, I don't think the message is a blocker, it can be updated in a future release. I'm just hoping we can get this into the current release 🤞

Comment (Humoiz) [reply azurerm/internal/services/resource/resource_arm_provider_registration.go:76]: I can make a label change very quick but I want @tombuildsstuff to agree with this too. Other than this if there is no other feedback than this is ready to be merged.

**Comment by Humoiz** [comment]
@katbyte @tombuildsstuff Any updates on this PR?

**Comment by jackofallops** [comment]
Hi @Humoiz It looks like there's a merge conflict holding this up as well as the test situation.  I think the testing needs to force a de-registration of a resource provider before the test starts in order to run/validate crrectly.

File: azurerm/internal/services/resource/resource_arm_provider_registration.go
Comment (Humoiz) [reply azurerm/internal/services/resource/resource_arm_provider_registration.go:76]: I want this PR to be merged soon now. This is very minor and I think can be updated outside of this PR.

**Comment by Humoiz** [comment]
> Hi @Humoiz It looks like there's a merge conflict holding this up as well as the test situation. I think the testing needs to force a de-registration of a resource provider before the test starts in order to run/validate crrectly.

@jackofallops - I have resolved merge conflicts and also updated the test to force de-registration of a resource provider. We do not have a terraform resource to de-register the resource provider so I used the provisioner with az cli command to unregister the resource.

**Comment by tombuildsstuff** [comment]
hey @Humoiz

Thanks for pushing those changes and apologies for the delayed re-review here!

I hope you don't mind but as a part of reviewing this, I've pulled this down locally, backed out the merges and squashed these commits down so that it's easier to rebase this on top of master.

I've pulled this resource down and tested this resource and this looked good - however I've ultimately ended up adding a couple of features to this:

* Enhanced Validation for the Resource Provider Name - this is validating the `name` exists at `terraform plan` time
* Validating the name of the Resource Provider at Import time, this ensures the casing is consistent with the Resource Provider Name
* Resource ID Formatter/Parser - this ensures the Resource ID is consistent and that somebody doesn't try and import a Resource Group ID as a Resource Provider ID (for example)
* Updating the Resource and Tests to be "Typed" - this is a new resource and so a good candidate for it since I was updating the other objects anyway

As mentioned above, I've squashed this down and rebased this PR on top of master - and the tests pass with the changes mentioned above:

```
$ TF_ACC=1 envchain azurerm go test -v ./azurerm/internal/services/resource/tests/ -run=TestAccResourceProviderRegistration_
=== RUN   TestAccResourceProviderRegistration_basic
=== PAUSE TestAccResourceProviderRegistration_basic
=== RUN   TestAccResourceProviderRegistration_requiresImport
=== PAUSE TestAccResourceProviderRegistration_requiresImport
=== CONT  TestAccResourceProviderRegistration_basic
=== CONT  TestAccResourceProviderRegistration_requiresImport
--- PASS: TestAccResourceProviderRegistration_basic (101.31s)
--- PASS: TestAccResourceProviderRegistration_requiresImport (172.01s)
PASS
ok  	github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/resource/tests	172.071s
```

Since I've pushed some broader changes here I'm going to ask @jackofallops to take another look through, but this should be good to go in the next release :thumbs_up: - apologies again for the delay getting this merged!

Thanks!

**Comment by Humoiz** [comment]
> hey @Humoiz
> 
> Thanks for pushing those changes and apologies for the delayed re-review here!
> 
> I hope you don't mind but as a part of reviewing this, I've pulled this down locally, backed out the merges and squashed these commits down so that it's easier to rebase this on top of master.
> 
> I've pulled this resource down and tested this resource and this looked good - however I've ultimately ended up adding a couple of features to this:
> 
> * Enhanced Validation for the Resource Provider Name - this is validating the `name` exists at `terraform plan` time
> * Validating the name of the Resource Provider at Import time, this ensures the casing is consistent with the Resource Provider Name
> * Resource ID Formatter/Parser - this ensures the Resource ID is consistent and that somebody doesn't try and import a Resource Group ID as a Resource Provider ID (for example)
> * Updating the Resource and Tests to be "Typed" - this is a new resource and so a good candidate for it since I was updating the other objects anyway
> 
> As mentioned above, I've squashed this down and rebased this PR on top of master - and the tests pass with the changes mentioned above:
> 
> ```
> $ TF_ACC=1 envchain azurerm go test -v ./azurerm/internal/services/resource/tests/ -run=TestAccResourceProviderRegistration_
> === RUN   TestAccResourceProviderRegistration_basic
> === PAUSE TestAccResourceProviderRegistration_basic
> === RUN   TestAccResourceProviderRegistration_requiresImport
> === PAUSE TestAccResourceProviderRegistration_requiresImport
> === CONT  TestAccResourceProviderRegistration_basic
> === CONT  TestAccResourceProviderRegistration_requiresImport
> --- PASS: TestAccResourceProviderRegistration_basic (101.31s)
> --- PASS: TestAccResourceProviderRegistration_requiresImport (172.01s)
> PASS
> ok  	github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/resource/tests	172.071s
> ```
> 
> Since I've pushed some broader changes here I'm going to ask @jackofallops to take another look through, but this should be good to go in the next release :thumbs_up: - apologies again for the delay getting this merged!
> 
> Thanks!

Looks good. Thank you!

**Review by jackofallops — APPROVED** [review]
LGTM 👍

**Pull request merged**
